Fix ftell() on Apple II to return the correct value.

ppelleti opened the pull request:
Fixes #722: `ftell()` returns the value returned by `lseek()`, and `lseek()` for the
Apple II wasn't returning a value.

ppelleti commented:
> How about an .ifdef __APPLE2ENH__ and stz?

OK, I've added that.

oliverschmidt commented:
> How about an .ifdef __APPLE2ENH__ and stz?

Using stz isn't primarily an attribute of the Apple II model, rather it's an attribute of the CPU inside the model. There's a predominant way to do that in the cc65 code base. E.g. like here: https://github.com/cc65/cc65/blob/master/libsrc/runtime/land.s

oliverschmidt commented:
__directerrno and __mappederrno return -1 as an 16 bit value. Here we want to return -1 as 32 bit value. So we need to set sreg/sreg+1 to $FF before jumping to them.

oliverschmidt commented on libsrc/apple2/lseek.s:
```diff
@@ -85,8 +85,12 @@ seek_common:
+        .else
         lda     #0
         sta     sreg+1
+        .endif
```
The Apple2 specific code uses `#$00` istead of `#0`. Please adjust the line above.

ppelleti commented:
OK, I believe I've fixed those three items.

oliverschmidt commented:
> OK, I believe I've fixed those three items.

Yes, you did - thanks for the very quick reaction :-)

Not directly related but maybe worth changing (as part of this pull request ?): The code as it stands ignores bits 25-32 of the offset parameter. Given that the code explicitly checks that whence is valid would it be consistent to check that bits 25-32 of the offset are actually zero - and just use the same exit used with a wrong whence when they don't?

ppelleti commented:
> would it be consistent to check that bits 25-32 of the offset are actually zero

I think that we would need to check that bits 25-32 are a sign extension of bit 24, since offset could be negative.

ppelleti commented:
I think what we need to do is check that bits 24-31 are 0, *after* adding `offset` to the 24-bit quantity in A, X, and Y in `seek_common`.

Sorry, I haven't been thinking very clearly at this time of night.  :)

oliverschmidt commented:
> [...] check that bits 24-31 are 0, after adding offset to the 24-bit quantity in A, X, and Y in seek_common.

So just saving X to ptr2+1 at the top and change the 3-byte-addition into a 4-byte-addition using ptr2+1 and afterwards check the 4th byte to be zero, correct?

That seems reasonable to me. Are you willing to implement and test that approach?

greg-king5 commented on libsrc/apple2/lseek.s:
```diff
@@ -84,13 +85,31 @@ seek_common:
+oserr:  ldx     #$FF
+        stx     sreg
+        stx     sreg+1
+        jmp     __mappederrno
```
A slightly better way:
```asm
oserr:  jsr     __mappederrno
        stx     sreg
        stx     sreg+1
        rts
```
It saves a byte.

ppelleti commented:
> That seems reasonable to me. Are you willing to implement and test that approach?

I've gone ahead and implemented that, and I implemented @greg-king5's suggestion to save a byte on each of the two error paths.

I'll let you know when I've finished testing.

ppelleti commented:
OK, I verified that `testcode/lib/seek.c` passes on my unenhanced Apple //e running ProDOS 1.9.  I also added some improvements to the test.

oliverschmidt commented:
Thanks for the contribution - and the feedback :-)
